### PR TITLE
ms-gsl: Fix CMake module file name, namespace, and target name for CMakeDeps generator

### LINKED ISSUE #7887: [package] ms-gsl/3.1.0: CMakeDeps + Conan 1.42 uses incorrect name for CMake configs
### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **ms-gsl/3.1.0**
  * Operating System+version: **Fedora Silverblue 35**
  * Compiler+version: **GCC 11**
  * Conan version: **conan 1.42.0**
  * Python version: **Python 3.10.0**

### Conan profile (output of `conan profile show default` or `conan profile show <profile>` if custom profile is in use)
```
[settings]
os=Linux
arch=x86_64
compiler=gcc
compiler.version=11
compiler.libcxx=libstdc++11
compiler.cppstd=17
build_type=Debug
[options]
[conf]
[build_requires]
[env]
```

### Steps to reproduce (Include if Applicable)

Use the `CMakeToolchain` / `CMakeDeps` generator for a project and add `ms-gsl` as a `requires`.
In your `CMakeLists.txt` use `find_package(Microsoft.GSL REQUIRED)` and it will not be found during CMake's configure stage.
This is because the package is now named `ms-gsl` instead, which is unexpected.

### Logs (Include/Attach if Applicable)
<details><summary>Click to expand log</summary>

```
-- Conan: Target declared 'fmt::fmt'
CMake Error at Module/Components/Platform/CMakeLists.txt:8 (find_package):
  By not providing "FindMicrosoft.GSL.cmake" in CMAKE_MODULE_PATH this
  project has asked CMake to find a package configuration file provided by
  "Microsoft.GSL", but CMake did not find one.

  Could not find a package configuration file provided by "Microsoft.GSL"
  with any of the following names:

    Microsoft.GSLConfig.cmake
    microsoft.gsl-config.cmake

  Add the installation prefix of "Microsoft.GSL" to CMAKE_PREFIX_PATH or set
  "Microsoft.GSL_DIR" to a directory containing one of the above files.  If
  "Microsoft.GSL" provides a separate development package or SDK, be sure it
  has been installed.


-- Configuring incomplete, errors occurred!
```

</details>

### PR DESCRIPTION
**Pull request by jwillikers**
Specify library name and version:  **ms-gsl/3.1.0**

Fixes #7887.
Corrects the config file name, namespace, and target name used by the `ms-gsl` target when using the CMakeDeps generator.
This appears to be necessary since Conan 1.42.

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by uilianries — APPROVED** [review]
LGTM

File: recipes/ms-gsl/all/conanfile.py [hunk 1]
```diff
@@ -70,6 +70,9 @@ def package_info(self):
         self.cpp_info.names["cmake_find_package_multi"] = "Microsoft.GSL"
         self.cpp_info.components["_ms-gsl"].names["cmake_find_package"] = "GSL"
         self.cpp_info.components["_ms-gsl"].names["cmake_find_package_multi"] = "GSL"
+        self.cpp_info.set_property("cmake_file_name", "Microsoft.GSL")
+        self.cpp_info.set_property("cmake_target_namespace", "Microsoft.GSL")
+        self.cpp_info.components["_ms-gsl"].set_property("cmake_target_name", "GSL")
```
Comment (SpaceIm) [inline recipes/ms-gsl/all/conanfile.py:75]: It's confusing to keep the old mechanism. I would remove it and bump `required_conan_version`

**Comment by SpaceIm** [comment]
Is not a regression in conan 1.42.0? I would have expected that CMakeDeps inherits from old mechanism until conan v2 is released. Now many CCI recipes are broken with CMakeDeps.

File: recipes/ms-gsl/all/conanfile.py (hunk 1 above)
Comment (jwillikers) [reply recipes/ms-gsl/all/conanfile.py:75]: I can do that.

**Comment by jwillikers** [comment]
> Is it not a regression in conan 1.42.0? I would have expected that CMakeDeps inherits from old mechanism until conan v2 is released. Now many CCI recipes are broken with CMakeDeps.

It might be, but I wasn't sure since CMakeDeps is still considered experimental. Honestly, I was hoping someone here would know if it's a regression or expected behavior.

File: recipes/ms-gsl/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/ms-gsl/all/conanfile.py:75]: Also you don't need anymore to use components I think:

```python
    def package_info(self):
        self.cpp_info.set_property("cmake_file_name", "Microsoft.GSL")
        self.cpp_info.set_property("cmake_target_namespace", "Microsoft.GSL")
        self.cpp_info.set_property("cmake_target_name", "GSL")
        if tools.Version(self.version) < "3.0.0":
            self.cpp_info.defines = [
                self._contract_map[str(self.options.on_contract_violation)]
            ]
```

Comment (jwillikers) [reply recipes/ms-gsl/all/conanfile.py:75]: Doesn't work with the test package. Do I have to modify the test package to use the `CMakeDeps` generator. Would this change not be backwards compatible with the other generators?

Comment (SpaceIm) [reply recipes/ms-gsl/all/conanfile.py]: It's important to still properly support cmake_find_package and cmake_find_package_multi in CCI recipes, so forget my last suggestion.

Comment (jwillikers) [reply recipes/ms-gsl/all/conanfile.py:75]: Okay, I've committed and reverted this change so its in the history, but for now I'm leaving the changes as they were with the original functions for the other generators. The component part `self.cpp_info.components["_ms-gsl"].set_property("cmake_target_name", "GSL")` is also necessary, or things won't build for the test package.

Comment (SpaceIm) [reply recipes/ms-gsl/all/conanfile.py:75]: Ok so I've tested with cmake_find_package, cmake_find_package_multi and CMakeDeps, here is a good trade-off 

```python
required_conan_version = ">=1.36.0"

(...)

    def package_info(self):
        self.cpp_info.set_property("cmake_file_name", "Microsoft.GSL")
        self.cpp_info.set_property("cmake_target_name", "Microsoft.GSL")
        self.cpp_info.components["_ms-gsl"].set_property("cmake_target_name", "GSL")
        if tools.Version(self.version) < "3.0.0":
            self.cpp_info.components["_ms-gsl"].defines = [
                self._contract_map[str(self.options.on_contract_violation)]
            ]
```

**Comment by SpaceIm** [comment]
see https://github.com/conan-io/conan-center-index/pull/7890#discussion_r739560974

File: recipes/ms-gsl/all/conanfile.py (hunk 1 above)
Comment (jwillikers) [reply recipes/ms-gsl/all/conanfile.py]: Works for me! Done.

**Review by SpaceIm — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**